### PR TITLE
Throw more descriptive exceptions when the constraint system is violated

### LINKED ISSUE #1668: logic error: Constraint system not satisfied by inputs
I tried sendmany rpc call like this.

```
"{"method":"z_sendmany","params":["ztaaga95QAPyp1kSQ1hD2kguCpzyMHjxWZqaYDEkzbvo7uYQYAw2S8X4Kx98AvhhofMtQL8PAXKHuZsmhRcanavKRKmdCzk",[{"address":"tmWYGUbthL7DrnQWszGNvTp7j2NUJpXGeoK","amount":0.0001},{"address":"ztScEWEw5SmyBLDs7awLES9pymXkXAGFkEymMyNhUYQn8TPCwdq6jAD6dYto8CMnYEKMeCMTkk9MNvLhqZP44CQgNx5223h","amount":0.0002}]],"id":48}"
```

And it shows like this when I request `zcash-cli z_getoperationstatus`

```
{
        "id" : "opid-1d2d3f7b-8a31-4f66-bcef-c6f6dad12d6b",
        "status" : "executing",
        "creation_time" : 1477592457
}
```

After some time I get like below.

```
{
        "id" : "opid-f2265594-44d8-4c66-8989-9238053c1194",
        "status" : "failed",
        "creation_time" : 1477591496,
        "error" : {
            "code" : -1,
            "message" : "logic error: Constraint system not satisfied by inputs"
        }
    }
```

This is my previous attempt. That's why operation id is not same.

Weird thing is that it actually succeeds time to time. I tried several times, some succeeded and some failed.
Is it related to this issue? https://github.com/zcash/zcash/blob/master/doc/security-warnings.md#block-chain-reorganization-major-differences

Seems like this take more than 5 minutes on my test pc, and block height changes between generating transaction. Not sure about it but just guessing.

### PR DESCRIPTION
**Pull request by ebfull**
Closes #1668.

### CODE REVIEW

**Comment by ebfull** [comment]
This is ready for review! :)

**Comment by str4d** [comment]
Reviewing.

File: src/gtest/test_joinsplit.cpp [hunk 1]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+        {
+            JSOutput(addr, 1900000000000000),
+            JSOutput(addr, 1900000000000000)
+        },
+        0,
+        0,
+        tree.root(),
+        "nonsensical right hand side of joinsplit balance");
+
+        // Absurd total output value
```
Comment (str4d) [inline src/gtest/test_joinsplit.cpp]: Copy-pasta comment.

File: src/wallet/gtest/test_wallet.cpp [hunk 2]
```diff
@@ -99,7 +99,7 @@ CWalletTx GetValidReceive(const libzcash::SpendingKey& sk, CAmount value, bool r
     // Prepare JoinSplits
     uint256 rt;
     JSDescription jsdesc {*params, mtx.joinSplitPubKey, rt,
-                          inputs, outputs, value, 0, false};
+                          inputs, outputs, 2*value, 0, false};
```
Comment (str4d) [inline src/wallet/gtest/test_wallet.cpp]: Is this (and subsequent changes in this file) for fixing issues that are now properly caught? If so, nice!

File: src/zcash/JoinSplit.cpp [hunk 3]
```diff
@@ -181,8 +182,44 @@ class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
...
+                }
+
+                // Ensure we have the key to this note.
+                if (inputs[i].note.a_pk != inputs[i].key.address().a_pk) {
+                    throw std::invalid_argument("input note not authorized to spend with given key");
+                }
+
+                // Balance must be sensical
+                if (inputs[i].note.value > MAX_MONEY) {
+                    throw std::invalid_argument("nonsensical input note value");
```
Comment (bitcartel) [inline src/zcash/JoinSplit.cpp]: input note value too big (greater than MAX_MONEY)

File: src/zcash/JoinSplit.cpp [hunk 4]
```diff
@@ -181,8 +182,44 @@ class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
...
+                }
+
+                lhs_value += inputs[i].note.value;
+
+                if (lhs_value > MAX_MONEY) {
+                    throw std::invalid_argument("nonsensical left hand size of joinsplit balance");
```
Comment (bitcartel) [inline src/zcash/JoinSplit.cpp]: input value too big (greater than MAX_MONEY)

File: src/zcash/JoinSplit.cpp [hunk 5]
```diff
@@ -197,12 +234,29 @@ class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
 
         // Compute notes for outputs
         for (size_t i = 0; i < NumOutputs; i++) {
+            // Sanity checks of output
+            {
+                if (outputs[i].value > MAX_MONEY) {
+                    throw std::invalid_argument("nonsensical output value");
```
Comment (bitcartel) [inline src/zcash/JoinSplit.cpp]: Ditto avove

File: src/zcash/JoinSplit.cpp [hunk 6]
```diff
@@ -181,8 +182,44 @@ class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
             throw std::runtime_error("JoinSplit proving key not loaded");
         }
 
-        // Compute nullifiers of inputs
+        if (vpub_old > MAX_MONEY) {
+            throw std::invalid_argument("nonsensical vpub_old value");
```
Comment (bitcartel) [inline src/zcash/JoinSplit.cpp]: vpub_old is invalid / too big, greater than MAX_MONEY

File: src/zcash/JoinSplit.cpp [hunk 7]
```diff
@@ -181,8 +182,44 @@ class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
...
+        }
+
+        if (vpub_new > MAX_MONEY) {
+            throw std::invalid_argument("nonsensical vpub_new value");
```
Comment (bitcartel) [inline src/zcash/JoinSplit.cpp]: ditto abive

File: src/gtest/test_joinsplit.cpp [hunk 8]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+        invokeAPIFailure(js,
+        {
+            JSInput(),
+            JSInput()
+        },
+        {
+            JSOutput(),
+            JSOutput()
+        },
+        2100000000000001,
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: How about MAX_MONEY + 1 ?

File: src/gtest/test_joinsplit.cpp [hunk 9]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+        {
+            JSInput(),
+            JSInput()
+        },
+        {
+            JSOutput(),
+            JSOutput()
+        },
+        0,
+        2100000000000001,
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: Ditto above

File: src/gtest/test_joinsplit.cpp [hunk 10]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+        tree.root(),
+        "nonsensical vpub_new value");
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: Ditto above

File: src/gtest/test_joinsplit.cpp [hunk 11]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+            JSInput()
+        },
+        {
+            JSOutput(),
+            JSOutput()
+        },
+        0,
+        0,
+        tree.root(),
+        "nonsensical input note value");
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: ditto above

File: src/gtest/test_joinsplit.cpp [hunk 12]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+        std::vector<ZCIncrementalWitness> witnesses;
+        ZCIncrementalMerkleTree tree;
+        increment_note_witnesses(uint256(), witnesses, tree);
+        SpendingKey sk = SpendingKey::random();
+        PaymentAddress addr = sk.address();
+        Note note1(addr.a_pk, 100, random_uint256(), random_uint256());
+        increment_note_witnesses(note1.cm(), witnesses, tree);
+        Note note2(addr.a_pk, 100, random_uint256(), random_uint256());
+        increment_note_witnesses(note2.cm(), witnesses, tree);
+        Note note3(addr.a_pk, 2100000000000001, random_uint256(), random_uint256());
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: MAX_MONEY + 1

File: src/gtest/test_joinsplit.cpp [hunk 13]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+            JSInput(witnesses[5], note5, sk)
+        },
+        {
+            JSOutput(),
+            JSOutput()
+        },
+        0,
+        0,
+        tree.root(),
+        "nonsensical left hand size of joinsplit balance");
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: ditto above

File: src/gtest/test_joinsplit.cpp [hunk 14]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+
+        // Absurd output value
+        invokeAPIFailure(js,
+        {
+            JSInput(),
+            JSInput()
+        },
+        {
+            JSOutput(addr, 2100000000000001),
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: MAX_MONEY + 1

File: src/gtest/test_joinsplit.cpp [hunk 15]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+            JSOutput()
+        },
+        0,
+        0,
+        tree.root(),
+        "nonsensical output value");
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: ditto above

File: src/gtest/test_joinsplit.cpp [hunk 16]
```diff
@@ -233,10 +290,204 @@ for test_input in TEST_VECTORS:
...
+            JSInput()
+        },
+        {
+            JSOutput(addr, 1900000000000000),
+            JSOutput(addr, 1900000000000000)
+        },
+        0,
+        0,
+        tree.root(),
+        "nonsensical right hand side of joinsplit balance");
```
Comment (bitcartel) [inline src/gtest/test_joinsplit.cpp]: > MAX_MONEY

**Comment by bitcartel** [comment]
ACK but see comments (which are none blocking)

File: src/wallet/gtest/test_wallet.cpp (hunk 2 above)
Comment (ebfull) [inline src/wallet/gtest/test_wallet.cpp]: Yes.

**Comment by ebfull** [comment]
[Here](https://github.com/ebfull/zcash/commit/75f28a936c1af22d3220fc7ab4585b3794877964) was my attempt before I realized there was an endless stream of globals to wade through.

**Comment by ebfull** [comment]
@zkbot r+

**Comment by daira** [comment]
Oh, ``GetJoinSplitValueIn`` sums the *outputs* of JoinSplits. That function name makes no sense to me (it should be something like ``GetTotalInputToTransparentPoolFromJoinSplits``), but I see how the fix works now.

**Comment by daira** [comment]
utACK

**Pull request merged**